### PR TITLE
Fixes in API and function visibility

### PR DESCRIPTION
**Pull request by traceon**
Review symbol visibility settings
Enable LTO/IPO where available for optimized builds
Removed dead/unused/unnecessary code pieces and legacy ODBC API function implementations
Fixed API macro usage
Added all the v3 function exports
Added reporting of actually implemented APIs in `SQLGetFunctions()`
Fixed some missing Unicode vs ANSI function declarations
Minimum CMake version is 3.9 now
Reporting `SQL_SUCCESS` on `SQLEndTran()` unconditionally

### CODE REVIEW

File: driver/api/impl/attr.cpp [hunk 1]
```diff
@@ -434,15 +434,19 @@ SQLRETURN GetStmtAttr(
 
 extern "C" {
 
+#if defined(__clang__) || defined(__GNUC__) || defined(__GNUG__)
+#   pragma GCC visibility push(default)
```
Comment (Enmk) [inline driver/api/impl/attr.cpp]: Super surprising that `SQL_API` symbols are not  explicitly exported/visible.

File: driver/api/impl/handles.cpp [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 #include <type_traits>
 
-namespace {
+namespace { namespace impl {
```
Comment (Enmk) [inline driver/api/impl/handles.cpp]: why do you need an additional namespace here? Aren't unnamed namespace are unique and non-visible?

File: driver/api/impl/attr.cpp (hunk 1 above)
Comment (traceon) [reply driver/api/impl/attr.cpp]: Looks like ODBC API conventions [unsurprisingly] came from pre-visibility, or even pre-dynamic libraries era. Specifying [public] visibility in the code causes "inconsistent linkage" when building in MSVC/MDAC environment.

File: driver/api/impl/handles.cpp (hunk 2 above)
Comment (traceon) [reply driver/api/impl/handles.cpp]: Just for readability/code organization: `impl::abc()` looks better than `impl_abc()`.

**Comment by traceon** [comment]
Yeah, that was my first approach, but since it was unsuitable for msvc/mdac, I used this one to introduce less diffs. No problem, I'll switch to your proposed approach. It won't be used for `_WIN32` case though: `X_API` will be empty for windows, and `.def` file will still be used. Another approach would be using pragma comments with linker flags fro msvc only. I was planning to switch to it from '.def''s in future iterations of api cleanups though.

**Pull request merged**